File: shared/src/main/java/com/emanuelvini/nuven/core/shared/player/profile/util/ProfileUtil.java
package com.emanuelvini.nuven.core.shared.player.profile.util;

import com.emanuelvini.nuven.core.shared.player.profile.Profile;
import com.emanuelvini.nuven.core.shared.player.ranking.Tag;
import lombok.val;

public class ProfileUtil {

    public static Profile from(String serialized) {
        val props = serialized.split(";");
        val owner = props[0];
        val cash = Double.parseDouble(props[1]);
        val tagProps = props[2].split(":");
        val tagOrder = Integer.parseInt(tagProps[0]);
        val tagName = tagProps[2];
        val tagTag = tagProps[3];
        val tagPermission = tagProps[4];

        return new Profile(
                owner,
                cash,
                new Tag(tagName, tagTag, tagOrder, tagPermission)
        );
    }

    public static String to(Profile profile) {

        val tag = profile.getTag();
        val cash = profile.getCash();
        val owner = profile.getOwner();

        return owner +
                ";" +
                cash +
                ";" +
                tag.getOrder() +
                ":" +
                tag.getName() +
                ":" +
                tag.getTag() +
                ":" +
                tag.getPermission() +
                ";";
    }

}

File: shared/src/main/kotlin/com/emanuelvini/nuven/core/shared/player/profile/util/ProfileUtil.kt
package com.emanuelvini.nuven.core.shared.player.profile.util

import com.emanuelvini.nuven.core.shared.player.profile.Profile
import com.emanuelvini.nuven.core.shared.player.ranking.Role


object ProfileUtil {
    fun from(serialized: String): Profile {
        val props = serialized.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val owner = props[0]
        val cash = props[1].toDouble()
        val roleProps = props[2].split(":")
        val roleOrder = roleProps[0].toInt()
        val rolePrefix = roleProps[1]
        val rolePermission = roleProps[2]
        val roleStaff = roleProps[3].toBoolean()
        val roleSpecialRole = roleProps[4].toBoolean()
        val roleJoinBroadcast = roleProps[5].toBoolean()
        val roleSpecialChat = roleProps[6].toBoolean()
        val roleName = roleProps[7]
        return Profile(
            owner,
            cash,
            Role(roleName, rolePrefix, rolePermission, roleOrder, roleStaff, roleSpecialRole, roleJoinBroadcast, roleSpecialChat)
        )
    }

    fun to(profile: Profile): String {
        val role = profile.role
        val cash = profile.cash
        val owner = profile.owner
        return "${owner};$cash;${role.order}:${role.prefix}:${role.permission}:${role.staffRole}:${role.specialRole}:${role.joinBroadcast}:${role.specialChat}:${role.name}"
    }
}
